crates/cache/src/oci.rs

```diff
@@ -0,0 +1,130 @@
+//! Cache for OCI registry entities.
+
+use anyhow::{Context, Result};
+use oci_distribution::Reference;
+use tokio::fs;
+
+use std::path::{Path, PathBuf};
+
+const CONFIG_DIR: &str = "fermyon";
+const REGISTRY_CACHE_DIR: &str = "registry";
+const OCI_CACHE_DIR: &str = "oci";
+const MANIFESTS_DIR: &str = "manifests";
+const WASM_DIR: &str = "wasm";
+const DATA_DIR: &str = "data";
+
+/// Cache for registry entities.
+pub struct Cache {
+    /// Root directory for the cache instance.
+    pub root: PathBuf,
+}
+
+impl Cache {
+    /// Create a new cache given an optional root directory.
+    pub async fn new(root: Option<PathBuf>) -> Result<Self> {
+        let root = match root {
+            Some(root) => root,
+            None => dirs::config_dir()
+                .context("cannot get configuration directory")?
+                .join(CONFIG_DIR),
+        };
+        let root = root.join(REGISTRY_CACHE_DIR).join(OCI_CACHE_DIR);
+        Self::ensure_dirs(&root).await?;
+
+        Ok(Self { root })
+    }
+
+    /// The manifests directory for the current cache.
+    pub fn manifests_dir(&self) -> PathBuf {
+        self.root.join(MANIFESTS_DIR)
+    }
+
+    /// The Wasm bytes directory for the current cache.
+    pub fn wasm_dir(&self) -> PathBuf {
+        self.root.join(WASM_DIR)
+    }
+
+    /// The data directory for the current cache.
+    pub fn data_dir(&self) -> PathBuf {
+        self.root.join(DATA_DIR)
+    }
+
+    /// Get the file path to a manifest given a reference.
+    /// If the directory for the manifest does not exist, this will create it.
+    pub async fn manifest_for_reference(&self, reference: &str) -> Result<PathBuf> {
```

Comment:
`reference: impl AsRef<str>`?

Reply:
...but why?

Reply:
Why not? Alternatively, why choose to be more strict in a public api? `impl AsRef<str>` has better caller ergonomics.